drivers/optiups.c

```diff
@@ -463,6 +484,27 @@ void upsdrv_updateinfo(void)
 	/* read some easy settings */
 	if ( optimodel == OPTIMODEL_ZINTO )
 		optifill( _pollv_zinto, sizeof(_pollv_zinto)/sizeof(_pollv_zinto[0]) );
+	else if ( optimodel == OPTIMODEL_PS ) {
+		optifill( _pollv_ps, sizeof(_pollv_ps)/sizeof(_pollv_ps[0]) );
+
+		r = optiquery( "NV" );
+		float inV = strtol ( _buf, NULL, 10);
```

Comment:
Does the call return an integer or a floating point number in the string? Why `strtol()` and not `long` but `float` to hold the result?

Reply:
Why strtol() and not long but float to hold the result?

I'll refer you to the "I don't work with C very often" comment in the PR. :) I just looked at the existing code and emulated what I saw.

The strings are integers.

Ir's quite late here, but I can sort it out in the morning. It's obviously not the right way to do things now you've pointed it out. :)

Reply:
So, looking at the code properly, neither _float_ nor _long_ seem warranted. Unless I'm missing something, a short integer from `atoi()` should be fine..? At least until someone plugs a UPS into a supply greater than 32kV, but I think if that happens a NUT driver is going to be the least of their worries. :)

As I say though, I'm not particularly fluent in C, so it's possible I've misunderstood something fundamental.

BTW, I'm not entirely clear on the difference between the `%i` and `%d` formats, both seemed to work. I went with `%d` arbitrarily.

Reply:
Well, it depends on strings returned and their character counts, mostly... I remember in some driver there was a problem of a protocol byte-sized voltage not sufficing for the 280V or so reality (which IIRC the LCD panel did reflect)...

Reply:
While here, it may make sense to use NUT's `str_to_*()` methods with error-checking, something #677/#966 stories were long about.

Reply:
As for the latest force-push - `atoi()` is specified as returning an `int`, so static checkers gotta complain about a `short int` there. Better safe than sorry :)

Reply:
Sorry if forced pushes caused any problems..? I keep forgetting to `--signoff`, so end up rebasing. :)

Using NUT's `str_to_*()` does make sense. Sorry, I would have done that earlier If I'd realised you had your own string library.

I've gone with `str_to_short()`. I wasn't clear if you were suggesting I should use `str_to_int()` specifically. The UPS returns a 4-character string for the voltages, with a leading zero, as such (from  `upsdrvctl -DDD`):

```
   4.715086     [D2] SEND: "NV"
   4.868114     [D2] READ: "0119"
   5.028188     [D2] SEND: "OV"
   5.187680     [D2] READ: "0119"
```

The raw output from a serial monitor in Windows 2000 running the official Opti-UPS software shows it split across two (or four) lines, but I don't know if that's relevant at all..?

```
 752	0.00147421	UPSCOMM.EXE	IRP_MJ_WRITE	Serial2	SUCCESS	Length 1: N	
 753	0.00197511	UPSCOMM.EXE	IRP_MJ_WRITE	Serial2	SUCCESS	Length 1: V	
 754	0.00194131	UPSCOMM.EXE	IRP_MJ_WRITE	Serial2	SUCCESS	Length 1: .	
 755	0.00481067	UPSCOMM.EXE	IRP_MJ_WRITE	Serial2	SUCCESS	Length 1: .	
 756	0.00000587	UPSCOMM.EXE	IRP_MJ_READ	Serial2	SUCCESS	Length 0: 	
 757	0.00005084	UPSCOMM.EXE	IRP_MJ_READ	Serial2	SUCCESS	Length 2: 01	
 758	0.00000615	UPSCOMM.EXE	IRP_MJ_READ	Serial2	SUCCESS	Length 3: 19.	
 759	0.00000587	UPSCOMM.EXE	IRP_MJ_READ	Serial2	SUCCESS	Length 1: .	
```

This is about where I hit my limit of really knowing what's going on The driver works with my UPS with `str_to_short()`. I'm unable to judge if I should expect it to have issues if a voltage hits "0256" or something, you're going to be a better judge of that than I am.

I'm happy to change it to an unsigned short or an integer, if you think either of those would be better.

(I went with a signed variable because I thought the UPS might actually want to send negative voltages if the power cable is dodgy and the live/neutral wires are swapped. I didn't want to make up a bad IEC cable to test that though. :))

Reply:
Hopefully, as long as it matches the method declaration, `short` should be good enough for this use-case (as long as C standards are followed on the platform and it is at least 16 bits wide).

Reply:
And yes, better not test it that way (with broken cable). I had a lot of sparks and one device less back in the day, when a computer was plugged into one wall outlet, its display into another, and the wall wires in the ancient building were attached to leads randomly...

Reply:
I don't even know if many people still run these old Opti-UPS devices..? So it's possible the use-case is quite narrow and I'm it. :)

In terms of platforms, btw, all my NUT testing has been on an Orange Pi Zero (as part of this work-in-progress: https://github.com/moonbuggy/Opti-UPS-expansion-slot-SBC).

So I can confirm it works on ARMv7 to '0240'. Which is really as high as it should ever get since that's the limit of the DIP switch setting and the reported live in/out voltages are half the real voltages. (I'd need an actual mains voltage of 480V to get anything higher than '0240' back from any of the three commands involved.)

Anyway, thanks for your help with the datatypes. :) Apologies if my limited C ability was annoying. :)

Reply:
Not at all, and thanks to you! :)